### ISSUE TITLE
Reduce bot cooldown

### ISSUE THREAD

**Issue by zer0k-z** [prestrafe/prestrafe-bot#2]
From my experience and other people's feedback 15 seconds seems to be really, really long. A 5-7 second cooldown should suffice.
I feel like it isn't likely to reach the API limit with this change. Also most of the new commands (if not all of them) are API independent, so it shouldn't create more problems.

Honestly, it would be nicer if we can get an api key for third party integrators, with increased limit but without POST permission... but maybe the technology isn't there yet.

**Comment by jangraefen** [comment]
You are correct on the API rate limit, just not on the API 😄. I was more worried about the [Twitch Chat API limits](https://dev.twitch.tv/docs/irc/guide#command--message-limits). It states that the limit is 20 messages in 30 seconds. My understand is, that this limit is not per channel, but for the whole bot.

The bot is currently serving 17 channels. It is very unlikely, that all of them stream at the same time, but 3 channels live seems like a fairly good worst case estimate. So with that in mind, lets have a look at different timeouts, assuming that the bot gets spammed in each channel for 30 seconds straight:

- 10 seconds: `3 * (30 / 10) = 3 * 3 = 9`
- 7 seconds: `3 * (30 / 7) = 3 * 4.3 = 12.9`
- 5 seconds: `3 * (30 / 5) = 3 * 6 = 18`
- 4 seconds: `3 * (30 / 4) = 3 * 7.5 =22.5`

So 5 seconds seems to be indeed the sweet spot and is a much better timeout. Would you want to submit a PR or should I have look some time?

**Comment by zer0k-z** [comment]
There's a slight issue here. Currently the cooldown seems to be per command, so you can use 20 different commands and take the bot down. I could be wrong though. So if we were going to make the cooldown 5 seconds, it should be across all commands.

One way to make sure the bot doesn't reach the limit is by forcing mod status on bot and check its status when they send a message. This makes the limit 100 instead of 20, this should be more than enough.

A possible alternative is to allow streamer to use their own bot to avoid the API limit, though this would require more work, and it would probably be a pain without a frontend website. Of course this would still get bottlenecked by GlobalAPI though, but no issue with Twitch API anymore!

Now one way to not get bottlenecked by the GlobalAPI is to query everything through the game server if they are global, that way we won't have to worry about reaching the limit, though it would require a bigger bandwidth.

You could potentially fill [this form](https://dev.twitch.tv/limit-increase/) to demand for a higher Twitch limit for the bot, but it seems to be on pause until next month.

**Comment by jangraefen** [comment]
I think a better way of approaching this would be to just implement the cooldown so that it tracks all channels instead of tracking on a per channel basis.

Everything else would require a more or less major redesign of the bot. Also getting a higher Twitch API quota would require me to fullfill additional requirements, as an upgrade to a verified bot account requires me to have automatic ways for making the bot join or leave your channel.

What do you think?

**Comment by zer0k-z** [comment]
For now, my idea is:

1. Remove the cooldown and keep track of the number of the messages sent by the bot globally, put requests that go over the limit in a "waiting room". 
2. Enforce bot moderator role in Twitch channels to get a much more comfortable limit of 100 messages over 30 seconds.

This shouldn't require too much reworks (at least I think so), and it should be sufficient for some time.

**Comment by jangraefen** [comment]
I don't like to enforce moderator priviliges though, as this is very intrusive and requires alot of trust by the streamer, which I don't want to require. I like options 1 more 🙂.

**Comment by zer0k-z** [comment]
Since the code is public, people would know exactly what the bot does. The bot clearly does not do anything else but responds in chat, and any mod-only actions should be easily seen in the log.

Both options are not mutually exclusive either, as we can check for mod permission whenever the bot wants to respond to chat. If the bot is not modded and the current message count is already over 15, put the message in the wait queue instead. Therefore people who trust the bots will be able to use the bots to its full potential.
